Feature/multiplayer things spawn options

UnreallyHard opened the pull request.

fabiangreffrath commented on src/net_defs.h:
```diff
@@ -210,6 +220,7 @@ typedef struct
 
     int num_players;
     int consoleplayer;
+    int mp_things_spawn_type; // [crispy]
```
Doesn't this change alone break multiplayer compatibility with Chocolate Doom?

UnreallyHard replied:
No, as far as I can tell. I launched a modified Crispy Doom server and an unmodified Chocolate Doom client, played a couple levels and didn't notice any sync problems. Since default mp_things_spawn_type equals 0 there is no consistency problems within net play. 

Of course, if a Crispy Doom server will use any mp_things_spawn_type > 0 there will be a consistency failure while using an unmodified Chocolate Doom client. However, same will happen if server will use dm3 rules.

UnreallyHard replied:
I moved new optional packet reads and writes to the bottom of the net methods so they wouldn't interfere with default properties.
And I checked consistency with both a Chocolate server and a Crispy client and vice versa. I didn't notice any sync problems or errors in the console.

> Doesn't this change alone break multiplayer compatibility with Chocolate Doom?

As far as I can tell, the only proper way to send multiplayer flags to a client is by using net_gamesettings_t struct and NET_WriteSettings and NET_ReadSettings functions. So changing mp_things_spawn_type is inevitable, I think. 
Unless we are to create new net_gamesettings_extended_t struct that will inherit net_gamesettings_t properties. But then we must change a lot of other methods to be able to use the new struct.

@fabiangreffrath Do you know, if there is another correct way to send multiplayer flags?

UnreallyHard replied:
I completely remade the feature, following your recommendations. Now this feature only:

- Adds single command line parameter (coop2)
- Sends this game option to a client using deathmatch parameter

Notes:
Using coop2 flag game doesn’t spawn any multiplayer things except for multiplayer monsters. This variant was chosen since the many people in doom community think that is the most faithful option. We don’t spawn any extra deathmatch pickups that change the balance of a coop game but we spawn all monster that were specially placed on the map for a coop game by the id Software. Anyway, if there is an interest, all mp spawn options can be brought back using deathmatch parameters bitmask values.

Since any deathmatch value greater than 0 means that it’s a deathmatch game, a lot of code changes would be made for `deathmatch = 4` to work as coop game. So instead, there is only a small modification that uses deathmatch original value in the net operations and uses `deathmatch = 0` and `coop2 = true` values for everything else.

fabiangreffrath commented:
Sorry, but before you put even more effort into this, it'd be only fair to tell you that I am not convinced of this concept and I am not going to merge it like this. The difference to DM3-support is that this added a single command line parameter and one or two more checks in the code. The approach here, however, adds a new command line parameter with different arguments and even modifies the netgame protocol. That's a no, sorry.

The one feature that's been asked for is "coop without additional weapons", so I'd suggest to implement this and either call it `-coop2` in analogy to `-dm3`, or `-no_coop_spawns` in analogy to `-coop_spawns` - but nothing more, please.

fabiangreffrath commented:
Please try to restrict the amount of changes to the ones in this PR:

https://github.com/fabiangreffrath/crispy-doom/commit/02720309aa67b6bfc54377ac24cb75dbe8a718af

UnreallyHard commented:
> Please try to restrict the amount of changes to the ones in this PR:
https://github.com/fabiangreffrath/crispy-doom/commit/02720309aa67b6bfc54377ac24cb75dbe8a718af

@fabiangreffrath I would like to make less changes but I don't think it's feasible to both keep code clean and do it.

**src/doom/d_net.c**
The main problem is that the code considers any deathmatch value greater than 0 to be a deathmatch game so it's applies deathmatch logic to the new value `deathmatch = 4.`  But that doesn't work in our case as we are making a new coop mode. I described this problem in the previous comment https://github.com/fabiangreffrath/crispy-doom/pull/1175#discussion_r1527099529
> Since any deathmatch value greater than 0 means that it’s a deathmatch game, a lot of code changes would be made for `deathmatch = 4` to work as coop game. So instead, there is only a small code changes that modify deathmatch value. Net operations continue to use a deathmatch original value and if `deathmatch == 4` then everything else uses `deathmatch = 0` and `coop2 = true` values.

So Because of all that, I made a change to src/doom/d_net.c to convert a net deathmatch value to a global var and vice versa. I moved conditions to the separate functions to make code cleaner.

**netgame_modes_t** 
I added `netgame_modes_t` typedef to fix the issue with code maintainability as new deathmatch mode means that developer needs to change maximum value in both src/net_common.c and src/setup/multiplayer.c. Furthermore, using constants instead of integers is much better in my opinion, since it's much easy to find changes introduced by the deathmatch mode.

**boolean coop2**
It was added to get rid of a deathmatch value dependency in the base code.

Please tell me, If you have any ideas how to make less changes and support high maintainability?
If having the less changes is the main priority then I guess netgame_modes_t typedef can be deleted in exchange for the less code maintainability and the less readability.

UnreallyHard commented:
> Please try to restrict the amount of changes to the ones in this PR:

@fabiangreffrath I additionally removed `netgame_modes_t enum` (to have less changes) in an exchange for the less code maintainability.

seascape commented:
Features suggested here would help make Crispy Doom's multimonitor support more usable in single-player... which is just what I was hoping to find.

Problem: To have "observer drones" showing one or more additional views of the gameworld you need to be in networked multiplayer, even though all Crispy clients may be local to your own computer. This creates disruptive differences from normal single-player: multi-only monster and weapon spans, and the player repawns into the same world state upon death (no death penalty). (Also can't easily load savegames, but that's another thing.)

The unwanted/different enemy/weapon spawns sound like they could be resolved with some of the changes suggested in this pull request, specifically
`None (3) - Nothing multiplayer related is spawned at all`

So that's one (unusual) reason this'd be a cool feature to have... it'd make multimonitor mode better for single-player.

(That said, I have also felt a desire for option 1 when doing co-op several times in the past:
`All except weapons (1) - The most often asked variant, ex. https://github.com/fabiangreffrath/crispy-doom/issues/681`
)

fabiangreffrath commented on src/doom/d_main.c:
```diff
@@ -108,6 +108,7 @@ boolean         nomonsters;	// checkparm of -nomonsters
 boolean         respawnparm;	// checkparm of -respawn
 boolean         fastparm;	// checkparm of -fast
 boolean         coop_spawns = false;	// [crispy] checkparm of -coop_spawns
+boolean         coop2 = false;	// [crispy] checkparm of -coop2
```
Do we èven need this? What if we simply track "coop2" as "deathmatch == 4"?
